src/app/modules/item/pages/group-progress-grid/group-progress-grid.component.ts

```diff
@@ -151,18 +150,24 @@ export class GroupProgressGridComponent implements OnChanges, OnDestroy {
     this.progressOverlay = undefined;
   }
 
+  refresh(): void {
+    this.refresh$.next();
+  }
+
   private setDataByBatch(data: Data, size = 25): void {
     const { rows } = data;
-    const maxCount = Math.ceil(rows.length / size);
+    const startIndex = this.data.rows.length;
+    const numberOfRowsToAppend = Math.max(rows.length - startIndex, 0);
+    const maxCount = Math.ceil(numberOfRowsToAppend / size);
+    this.data = { ...data, rows: rows.slice(0, startIndex) };
 
-    this.data = { ...data, rows: [] };
     animationFrames().pipe(
       takeUntil(this.destroy$),
       scan(count => count + 1, 0),
       takeWhile(count => count <= maxCount),
       map(count => count * size),
     ).subscribe(lastIndex => {
-      this.data.rows = rows.slice(0, lastIndex);
+      this.data.rows = rows.slice(0, startIndex + lastIndex);
     });
```

Comment:
Will it work if we refresh faster than your function?  
I think it would be useful that your `setDataByBatch` is a rx pipe which is fill the data into `this.data` in steps instead of doing this operation "outside the flow".

Actually it would even cleanify the ugly `this.state = state.tag;  if (state.isReady) this.setDataByBatch(state.data);`. And modify building `this.data` based on a call in subscribe while `dataFetching` may change at any moment, while the fetching state also has data, ... it is sure there are case in which it will fail if you are not doing a switch map 🙄

Reply:
The switch map occurs in the component, this function is called in the subscribe callback. If there are two quick calls, the first will ignored in favour of the second.
As for the refresh button, I'm not _sure_ because it relies on the `mapToFetchState({ resetter })`, but since the source is given again, I guess it also contains the switch map. Plus, when clicking refresh, the list is greyed out, a user might click twice if s⋅he is ultra quick, otherwise it will be limited to one click.

I can try doing it in RxJS, but I doubt it will change anything about the refresh issue.

Reply:
Fixed

Reply:
Fixed, with tests